design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,255 @@
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+    logs_collections:
+      - name: basic-input
+        type: basics
+  - name: file-output
+    type: files
+    logs_collections:
+      - name: file-input
+        type: files
+```
+In the above example, the inputs from imjournal (basic-input) are forwarded to the remote host (forward-output); the inputs from imfile (file-input) are stored in the local file (file-output).
+
+With this style, it is not straightforward to implement one input to send multiple outputs in the logging roles.
+Moreover, to configure the input from `imjournal` to send both the remote host and the local file,
+the basics input would be placed under both forwards and files outputs as follows.
+It could give an impression basic-input0 and basic-input1 could be configured independently, but it is not.
+There is only one `imjournal` input and most likely the first one is wiped out.
+If possible, we should avoid this type of confusions in the designing phase.
+
+**Previous style example with uncertainty**
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+    logs_collections:
+      - name: basic-input0
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 1000
+        rsyslog_imjournal_ratelimit_interval: 10
+        rsyslog_imjournal_ratelimit_burst: 100000
+  - name: file-output
+    type: files
+    logs_collections:
+      - name: basic-input1
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 3000
+        rsyslog_imjournal_ratelimit_interval: 30
+        rsyslog_imjournal_ratelimit_burst: 300000
+```
+
+### logging_outputs, logs_collections and rsyslog_use_OUTPUT_ruleset
+This is a typical example of the configuration.
+In logging_outputs, the to-be-configured output roles are listed.
+In logs_collections, the to-be-configure input roles are listed.
+Each input role has `rsyslog_use_OUTPUT_ruleset: true|false` to specify which output the log messages are sent to.
+```
+logging_outputs:
+  - name: forward-output
+    type: forwards
+    rsyslog_forwards_actions:
+      - name: to-remote
+        protocol: tcp
+        target: remote_host_name.remote_domain_name
+        port: 514
+  - name: file-output
+    type: files
+logs_collections:
+  - name: basic-input
+    type: basics
+    rsyslog_use_files_ruleset: true
+  - name: file-input
+    type: files
+    rsyslog_use_files_ruleset: true
+    rsyslog_use_forwards_ruleset: true
+```
+There are some exceptions.
+- The files output is configured, by default. That is, even if there is no input roles that write logs to the local files, the files output config file is deployed. The filters and the actions in the configuration file (30-output.files) are equivalent to the ones in the original rsyslog.conf.
```

Comment:
What does it mean "by default"? Does it mean that if you don't supply any `logging_outputs`, there will be a default one consisting of one files output? (That's what I understand by "default".) Or, do you mean that a files output will be always appended to user-supplied `logging_outputs`, if it does not have any?

Reply:
> What does it mean "by default"? Does it mean that if you don't supply any `logging_outputs`, there will be a default one consisting of one files output? (That's what I understand by "default".) 

Correct.

> Or, do you mean that a files output will be always appended to user-supplied `logging_outputs`, if it does not have any?

That's correct, too.  The files output is always appended.  I thought it provides the same functionality as the current rsyslog_default does - deploying the original rsyslog.conf.

Reply:
What about having the default only for situations where the user has not supplied any `logging_outputs`? It is quite irregular to have a magic value that gets always appended.

Reply:
That's a problem with default values for lists in Ansible. it is not easy to have a list with some default values and then append to the list instead of overriding it.

Reply:
> What about having the default only for situations where the user has not supplied any `logging_outputs`? It is quite irregular to have a magic value that gets always appended.

Sorry for my repeating this example...  This is the simplest configuration (imjournal + omfile).  We need some input info (in this case imjournal), but we'd like to allow to specify output to the /var/log/messages, etc. since it's too obvious for the UNIX users...
```
logging_inputs:
   - name: basics
```
> That's a problem with default values for lists in Ansible. it is not easy to have a list with some default values and then append to the list instead of overriding it.

This is going to be identical to the above.  Do you think it's a bad idea?
```
logging_outputs:
   - name: files
logging_inputs:
   - name: basics
```